symengine/serialize-cereal.h

```diff
@@ -25,6 +25,113 @@
 namespace SymEngine
 {
 
+template <class Archive>
+class RCPBasicAwareOutputArchive : public Archive
+{
+    using Archive::Archive;
+
+public:
+    void save_rcp_basic(const RCP<const Basic> &ptr)
+    {
+        uintptr_t addr = (uintptr_t)(void *)ptr.get();
+        (*this)(addr);
+
+        auto id = _addresses.find(addr);
+        bool first_seen = (id == _addresses.end());
+        (*this)(first_seen);
+
+        if (not first_seen) {
+            return;
+        }
+        TypeID type_code = ptr->get_type_code();
+        save_typeid(*this, type_code);
+        switch (type_code) {
+#define SYMENGINE_ENUM(type, Class)                                            \
+    case type:                                                                 \
+        save_basic(*this, static_cast<const Class &>(*ptr));                   \
+        break;
+#include "symengine/type_codes.inc"
+#undef SYMENGINE_ENUM
+            default:
+                save_basic(*this, *ptr);
+        }
+        _addresses.insert(addr);
+    }
+
+private:
+    std::set<uintptr_t> _addresses;
+    //! Overload the rtti function to enable dynamic_cast
+    void rtti(){};
```

Comment:
I can see that this is a customization point for cereal:
https://github.com/USCiLab/cereal/blob/d1fcec807b372f04e4c1041b3058e11c12853e6e/include/cereal/details/helpers.hpp#L266

I'm just curious, I see that they already declare the destructor to be virtual there, probably my c++-knowledge that is failing here, but I thought that was enough to make our subclass here polymorphic? Or do we need such an overload here to ensure this?

Anyhow, would it be instructive to add an `override` specifier here for enhanced readability?

Reply:
I was following https://github.com/USCiLab/cereal/blob/d1fcec807b372f04e4c1041b3058e11c12853e6e/include/cereal/archives/adapters.hpp#L126-L127. Maybe it's not needed. I'll try.